<?xml version="1.0" encoding="us-ascii"?>
<html><head><title>SNPViewer Enhancements</title>
<meta content="text/html; charset=us-ascii" http-equiv="Content-Type" /><link rel="stylesheet" type="text/css" href="../Theme/nav.css" /><link rel="stylesheet" type="text/css" href="../Theme/popupmenu.css" /><link rel="stylesheet" type="text/css" href="../Theme/lf.css" /><link rel="stylesheet" type="text/css" href="../LinksExt/d2hTarget.css" />
<script type="text/javascript">
//<![CDATA[



var _needIndentation = false;




//]]></script>

<script language="JavaScript" src="../linker.js" type="text/JavaScript"></script>

<script language="JavaScript" src="../special.js" type="text/JavaScript"></script>

<script language="JavaScript" src="../common.js" type="text/JavaScript"></script>

<meta name="Generator" content="ComponentOne Doc-To-Help v.2011.1.0.614" />
<meta name="D2HTheme" content="(Default) v.146" />
</head>
<body style="PADDING-BOTTOM: 0px; BACKGROUND-COLOR: #535353; MARGIN: 7px; PADDING-LEFT: 0px; PADDING-RIGHT: 0px; OVERFLOW: hidden; PADDING-TOP: 0px" onload="d2hOnLoadToolbar(event);" relpart="../">
<script language="javascript" type="text/javascript">
//<![CDATA[



d2hTopicPreOpen();



d2hTrySyncTOC();







var prev = true;





var next = true;



if (typeof d2hInitMainThemeHandlers != "undefined")

    d2hInitMainThemeHandlers(prev, next);

d2hAdjustTopicBorders(document);    



if (typeof d2hLoadWindows != "undefined")

    d2hLoadWindows();

if (typeof d2hLoadNavUrls != "undefined")

    d2hLoadNavUrls();




//]]></script>

<div style="PADDING-BOTTOM: 0px; BACKGROUND-COLOR: white; MARGIN: 0px; PADDING-LEFT: 0px; WIDTH: 100%; PADDING-RIGHT: 0px; HEIGHT: 100%; OVERFLOW: auto; PADDING-TOP: 0px" id="D2HTopicOuterBody" class="d2h_outer_topic">
<div id="nsbanner" class="d2hNavigatorTopic">
<div style="PADDING-BOTTOM: 0px; PADDING-LEFT: 0px; PADDING-RIGHT: 0px; PADDING-TOP: 0px" id="TitleRow" class="d2h_main_titlerow">
<div style="DISPLAY: none"><a id="D2HPreviousAnchor" href="sampleexplorer.htm"></a><a id="D2HNextAnchor" href="integratedgenomicsviewer.htm"></a></div></div>
<script language="javascript" type="text/javascript">
//<![CDATA[

if (typeof d2hProcessTopicNavForCSH != "undefined")

    d2hProcessTopicNavForCSH();


//]]></script>
</div>
<div style="PADDING-BOTTOM: 10px; PADDING-LEFT: 22px; PADDING-RIGHT: 10px; HEIGHT: 100%; PADDING-TOP: 5px" id="nstext" class="d2h_main_nstext" valign="bottom">
<div id="d2h_breadcrumbs" class="d2h_breadcrumbs"><span class="d2h_breadcrumbsLabel"><b><span style="COLOR: #000080">Release 
Notes</span></b>&#160;&#160;</span><a class="d2h_breadcrumbsNormal" href="transmartreleasenotes.htm">tranSMART Release Notes</a><span class="d2h_breadcrumbsLinkSeparator"> &gt; </span><span class="d2h_breadcrumbsContentsOnly">New Features in This Release</span></div>
<h4><a name="_Toc297145010" id="_Toc297145010">SNPViewer Enhancements</a></h4>
<p class="MsoNormal">Enhancements that allow you to filter SNP array data multiple 
ways have been introduced in this release. You were limited in the last release 
to select one or more chromosomes to include in the visualization. In addition 
to filtering by chromosome, you are now able to filter by gene, SNP rs ID, or a 
combination of genes and SNP rs IDs. The complete list of filtering methods is 
described below:</p>
<table style="BORDER-BOTTOM: medium none; BORDER-LEFT: medium none; BORDER-COLLAPSE: collapse; MARGIN-LEFT: 5.4pt; BORDER-TOP: medium none; BORDER-RIGHT: medium none" class="MsoTableGrid" border="1" cellspacing="0" cellpadding="0">
  <thead>
  <tr style="PAGE-BREAK-INSIDE: avoid">
    <td style="BORDER-BOTTOM: windowtext 1.5pt double; BORDER-LEFT: black 1.5pt solid; PADDING-BOTTOM: 0pt; PADDING-LEFT: 5.4pt; WIDTH: 207.8pt; PADDING-RIGHT: 5.4pt; BORDER-TOP: black 1.5pt solid; BORDER-RIGHT: black 1pt solid; PADDING-TOP: 0pt" valign="top" width="277">
      <p class="TableHeading0">Filtering Method</p></td>
    <td style="BORDER-BOTTOM: windowtext 1.5pt double; BORDER-LEFT: medium none; PADDING-BOTTOM: 0pt; PADDING-LEFT: 5.4pt; PADDING-RIGHT: 5.4pt; BORDER-TOP: black 1.5pt solid; BORDER-RIGHT: black 1.5pt solid; PADDING-TOP: 0pt" valign="top">
      <p class="TableHeading0">Description</p></td></tr></thead>
  <tbody>
  <tr style="PAGE-BREAK-INSIDE: avoid">
    <td style="BORDER-BOTTOM: black 1pt solid; BORDER-LEFT: black 1.5pt solid; PADDING-BOTTOM: 0pt; PADDING-LEFT: 5.4pt; WIDTH: 207.8pt; PADDING-RIGHT: 5.4pt; BORDER-TOP: medium none; BORDER-RIGHT: black 1pt solid; PADDING-TOP: 0pt" valign="top" width="277">
      <p class="TableText0">By Chromosome</p></td>
    <td style="BORDER-BOTTOM: black 1pt solid; BORDER-LEFT: medium none; PADDING-BOTTOM: 0pt; PADDING-LEFT: 5.4pt; PADDING-RIGHT: 5.4pt; BORDER-TOP: medium none; BORDER-RIGHT: black 1.5pt solid; PADDING-TOP: 0pt" valign="top">
      <p class="TableText0">Select one or more chromosomes to include in the 
      visualization, or select <b>ALL</b> to include all chromosomes. If gene or 
      SNP rs IDs are entered, the selection of chromosomes is ignored.</p></td></tr>
  <tr style="PAGE-BREAK-INSIDE: avoid">
    <td style="BORDER-BOTTOM: black 1pt solid; BORDER-LEFT: black 1.5pt solid; PADDING-BOTTOM: 0pt; PADDING-LEFT: 5.4pt; WIDTH: 207.8pt; PADDING-RIGHT: 5.4pt; BORDER-TOP: medium none; BORDER-RIGHT: black 1pt solid; PADDING-TOP: 0pt" valign="top" width="277">
      <p class="TableText0">By Gene</p></td>
    <td style="BORDER-BOTTOM: black 1pt solid; BORDER-LEFT: medium none; PADDING-BOTTOM: 0pt; PADDING-LEFT: 5.4pt; PADDING-RIGHT: 5.4pt; BORDER-TOP: medium none; BORDER-RIGHT: black 1.5pt solid; PADDING-TOP: 0pt" valign="top">
      <p class="TableText0">To select a gene, type all or part of the gene into 
      the <b>Add a Gene</b> field.</p>
      <p class="TableText0">Separate multiple genes with commas.</p>
      <p class="TableText0"><span class="Bold"><span style="COLOR: #1f497d">Note:</span></span>&#160;&#160; Some genes that you 
      select as filters in tranSMART may not be found in the GenePattern 
      SNPViewer.</p></td></tr>
  <tr style="PAGE-BREAK-INSIDE: avoid">
    <td style="BORDER-BOTTOM: black 1pt solid; BORDER-LEFT: black 1.5pt solid; PADDING-BOTTOM: 0pt; PADDING-LEFT: 5.4pt; WIDTH: 207.8pt; PADDING-RIGHT: 5.4pt; BORDER-TOP: medium none; BORDER-RIGHT: black 1pt solid; PADDING-TOP: 0pt" valign="top" width="277">
      <p style="PAGE-BREAK-AFTER: avoid" class="TableText0">By SNP rs ID</p></td>
    <td style="BORDER-BOTTOM: black 1pt solid; BORDER-LEFT: medium none; PADDING-BOTTOM: 0pt; PADDING-LEFT: 5.4pt; PADDING-RIGHT: 5.4pt; BORDER-TOP: medium none; BORDER-RIGHT: black 1.5pt solid; PADDING-TOP: 0pt" valign="top">
      <p style="PAGE-BREAK-AFTER: avoid" class="TableText0">To add a SNP rs ID, 
      type the full ID into the <b>Selected SNPs</b> field. Separate multiple 
      IDs with commas.</p></td></tr>
  <tr style="PAGE-BREAK-INSIDE: avoid">
    <td style="BORDER-BOTTOM: black 1.5pt solid; BORDER-LEFT: black 1.5pt solid; PADDING-BOTTOM: 0pt; PADDING-LEFT: 5.4pt; WIDTH: 207.8pt; PADDING-RIGHT: 5.4pt; BORDER-TOP: medium none; BORDER-RIGHT: black 1pt solid; PADDING-TOP: 0pt" valign="top" width="277">
      <p class="TableText0">By a combination of genes or SNP rs IDs</p></td>
    <td style="BORDER-BOTTOM: black 1.5pt solid; BORDER-LEFT: medium none; PADDING-BOTTOM: 0pt; PADDING-LEFT: 5.4pt; PADDING-RIGHT: 5.4pt; BORDER-TOP: medium none; BORDER-RIGHT: black 1.5pt solid; PADDING-TOP: 0pt" valign="top">
      <p class="TableText0">Follow the instructions above to add a combination of 
      genes or SNP rs IDs.&#160; </p></td></tr></tbody></table>
<p class="MsoNormal">&#160;</p></div></div></body></html>
