<?xml version="1.0" encoding="us-ascii"?>
<html><head><title>SNP Viewer</title>
<meta content="text/html; charset=us-ascii" http-equiv="Content-Type" /><link rel="stylesheet" type="text/css" href="../Theme/nav.css" /><link rel="stylesheet" type="text/css" href="../Theme/popupmenu.css" /><link rel="stylesheet" type="text/css" href="../Theme/lf.css" /><link rel="stylesheet" type="text/css" href="../LinksExt/d2hTarget.css" /><link rel="stylesheet" type="text/css" href="../LinksExt/tranSMART_RN.css" />
<script type="text/javascript">
//<![CDATA[



var _needIndentation = false;




//]]></script>

<script language="JavaScript" src="../linker.js" type="text/JavaScript"></script>

<script language="JavaScript" src="../special.js" type="text/JavaScript"></script>

<script language="JavaScript" src="../common.js" type="text/JavaScript"></script>

<meta name="Generator" content="ComponentOne Doc-To-Help v.2011.1.0.614" />
<meta name="D2HTheme" content="(Default) v.146" />
</head>
<body style="PADDING-BOTTOM: 0px; BACKGROUND-COLOR: #535353; MARGIN: 7px; PADDING-LEFT: 0px; PADDING-RIGHT: 0px; OVERFLOW: hidden; PADDING-TOP: 0px" onload="d2hOnLoadToolbar(event);" relpart="../">
<script language="javascript" type="text/javascript">
//<![CDATA[



d2hTopicPreOpen();



d2hTrySyncTOC();







var prev = true;





var next = true;



if (typeof d2hInitMainThemeHandlers != "undefined")

    d2hInitMainThemeHandlers(prev, next);

d2hAdjustTopicBorders(document);    



if (typeof d2hLoadWindows != "undefined")

    d2hLoadWindows();

if (typeof d2hLoadNavUrls != "undefined")

    d2hLoadNavUrls();




//]]></script>

<div style="PADDING-BOTTOM: 0px; BACKGROUND-COLOR: white; MARGIN: 0px; PADDING-LEFT: 0px; WIDTH: 100%; PADDING-RIGHT: 0px; HEIGHT: 100%; OVERFLOW: auto; PADDING-TOP: 0px" id="D2HTopicOuterBody" class="d2h_outer_topic">
<div id="nsbanner" class="d2hNavigatorTopic">
<div style="PADDING-BOTTOM: 0px; PADDING-LEFT: 0px; PADDING-RIGHT: 0px; PADDING-TOP: 0px" id="TitleRow" class="d2h_main_titlerow">
<div style="DISPLAY: none"><a id="D2HPreviousAnchor" href="prerequisites.htm"></a><a id="D2HNextAnchor" href="heatmapperformanceenhancements.htm"></a></div></div>
<script language="javascript" type="text/javascript">
//<![CDATA[

if (typeof d2hProcessTopicNavForCSH != "undefined")

    d2hProcessTopicNavForCSH();


//]]></script>
</div>
<div style="PADDING-BOTTOM: 10px; PADDING-LEFT: 22px; PADDING-RIGHT: 10px; HEIGHT: 100%; PADDING-TOP: 5px" id="nstext" class="d2h_main_nstext" valign="bottom">
<div id="d2h_breadcrumbs" class="d2h_breadcrumbs"><span class="d2h_breadcrumbsLabel"><b><span style="COLOR: #000080">Release 
Notes</span></b>&#160;&#160;</span><a class="d2h_breadcrumbsNormal" href="transmartreleasenotes.htm">tranSMART Release Notes</a><span class="d2h_breadcrumbsLinkSeparator"> &gt; </span><span class="d2h_breadcrumbsContentsOnly">New Features in This Release</span></div>
<h4><a name="_Toc242498268" id="_Toc242498268"></a><a name="_Toc287175943" id="_Toc287175943">SNP Viewer</a></h4>
<p class="MsoNormal">The SNP Viewer allows you to analyze individual base 
variations in DNA sequences for normal and tumor tissue samples in a SNP 
array.&#160; The viewer supports both copy number analysis and loss of 
heterozygosity (LOH) analysis.</p>
<p class="MsoNormal">SNP array data, such as Affymetrix Genome-Wide Human SNP 
Array, is required for the SNP Viewer.&#160; In this release, the only study 
that contains SNP array data is the public study GSE19539 on ovarian 
cancer.<span class="MsoHyperlink"><span style="FONT-SIZE: 9pt"></span></span></p>
<p class="MsoNormal">Dataset Explorer uses the Broad Institute&#8217;s GenePattern 
genomic analysis platform to generate SNP visualizations.&#160; For information 
about the Broad Institute&#8217;s SNPViewer, see the document <span class="Bold">SNPViewer </span>on the following site:</p>
<p class="MsoNormal"><a href="http://www.broadinstitute.org/cgi-bin/cancer/software/genepattern/modules/gp_modules.cgi"><span style="FONT-SIZE: 9pt">http://www.broadinstitute.org/cgi-bin/cancer/software/genepattern/modules/gp_modules.cgi</span></a></p>
<p style="tab-stops: 0pt" class="ListNumStart">To run the SNP Viewer:</p>
<p style="tab-stops: 18.0pt 36.0pt 54.0pt 72.0pt 90.0pt 108.0pt" class="MsoListNumber">1.&#160; Run tranSMART, then click the <span class="Bold">Dataset Explorer</span> tab.</p>
<p style="tab-stops: 18.0pt 36.0pt 54.0pt 72.0pt 90.0pt 108.0pt" class="MsoListNumber">2.&#160; Select the study of of interest.</p>
<p style="tab-stops: 18.0pt 36.0pt 54.0pt 72.0pt 90.0pt 108.0pt" class="MsoListNumber">3.&#160; Drag the SNP array data into an empty box in one or 
both subsets.</p>
<p style="tab-stops: 18.0pt 36.0pt 54.0pt 72.0pt 90.0pt 108.0pt" class="MsoListNumber">4.&#160; Define the cohorts whose data points will be 
represented in the SNP visualization.</p>
<p style="tab-stops: 18.0pt 36.0pt 54.0pt 72.0pt 90.0pt 108.0pt" class="MsoListNumber">5.&#160; Click <span class="Bold">Advanced &gt; 
SNPViewer</span>.</p>
<p style="tab-stops: 18.0pt 36.0pt 54.0pt 72.0pt 90.0pt 108.0pt" class="MsoListNumber">6.&#160; On the SNPViewer dialog, select one or more 
chromosomes to include in the visualization, or select <span class="Bold">ALL</span> to include all chromosomes.</p>
<p style="MARGIN-RIGHT: 24pt" class="Note"><b><span style="COLOR: #1f497d">Note:</span></b><span style="COLOR: black">&#160; For 
performance reasons, select only those chromosomes that are of particular 
interest to you.&#160; The dataset for all chromosomes is extremely large.&#160; 
</span></p>
<p style="tab-stops: 18.0pt 36.0pt 54.0pt 72.0pt 90.0pt 108.0pt" class="MsoListNumber">7.&#160; Click <span class="Bold">OK</span> to display the 
visualization.</p>
<p class="MsoNormalIndent"><span style="COLOR: black">The visualization may 
require several minutes to display a large dataset, such as when multiple 
chromosomes are selected.</span></p></div></div></body></html>
