<!DOCTYPE html>
<HTML><HEAD><TITLE>SNP Viewer</TITLE>
<META content="text/html; charset=us-ascii" http-equiv=Content-Type><LINK 
rel=stylesheet type=text/css href="../Theme/nav.css"><LINK rel=stylesheet 
type=text/css href="../Theme/popupmenu.css"><LINK rel=stylesheet type=text/css 
href="../Theme/lf.css"><LINK rel=stylesheet type=text/css 
href="../LinksExt/doc2help.css"><LINK rel=stylesheet type=text/css 
href="../LinksExt/tranSMART_RN.css">
<SCRIPT type=text/javascript>



var _needIndentation = false;



</SCRIPT>

<SCRIPT language=JavaScript src="../linker.js">

</SCRIPT>

<SCRIPT language=JavaScript src="../special.js">

</SCRIPT>

<SCRIPT language=JavaScript src="../common.js">

</SCRIPT>

<meta name="Generator" content="ComponentOne Doc-To-Help v.2011.1.0.614">
<meta name="D2HTheme" content="(Default) v.146">
</HEAD>
<BODY 
style="PADDING-BOTTOM: 0px; BACKGROUND-COLOR: #535353; MARGIN: 7px; PADDING-LEFT: 0px; PADDING-RIGHT: 0px; OVERFLOW: hidden; PADDING-TOP: 0px" 
onload=d2hOnLoadToolbar(event); relPart="../">
<SCRIPT language=javascript>



d2hTopicPreOpen();



d2hTrySyncTOC();







var prev = true;





var next = true;



if (typeof d2hInitMainThemeHandlers != "undefined")

    d2hInitMainThemeHandlers(prev, next);

d2hAdjustTopicBorders(document);    



if (typeof d2hLoadWindows != "undefined")

    d2hLoadWindows();

if (typeof d2hLoadNavUrls != "undefined")

    d2hLoadNavUrls();



</SCRIPT>

<DIV 
style="PADDING-BOTTOM: 0px; BACKGROUND-COLOR: white; MARGIN: 0px; PADDING-LEFT: 0px; WIDTH: 100%; PADDING-RIGHT: 0px; HEIGHT: 100%; OVERFLOW: auto; PADDING-TOP: 0px" 
id=D2HTopicOuterBody class=d2h_outer_topic>
<DIV id=nsbanner class=d2hNavigatorTopic>
<DIV 
style="PADDING-BOTTOM: 0px; PADDING-LEFT: 0px; PADDING-RIGHT: 0px; PADDING-TOP: 0px" 
id=TitleRow class=d2h_main_titlerow>
<DIV style="DISPLAY: none"><A id=D2HPreviousAnchor 
href="prerequisites.htm"></A><A id=D2HNextAnchor 
href="onlinehelp.htm"></A></DIV></DIV>
<SCRIPT language=javascript>

if (typeof d2hProcessTopicNavForCSH != "undefined")

    d2hProcessTopicNavForCSH();

</SCRIPT>
</DIV>
<DIV 
style="PADDING-BOTTOM: 10px; PADDING-LEFT: 22px; PADDING-RIGHT: 10px; HEIGHT: 100%; PADDING-TOP: 5px" 
id=nstext class=d2h_main_nstext valign="bottom">
<DIV id=d2h_breadcrumbs class=d2h_breadcrumbs><SPAN 
class=d2h_breadcrumbsLabel><B><SPAN style="COLOR: #000080">Release 
Notes</SPAN></B>&nbsp;&nbsp;</SPAN><A class=d2h_breadcrumbsNormal 
href="transmartreleasenotes.htm">tranSMART Release Notes</A><SPAN 
class=d2h_breadcrumbsLinkSeparator> &gt; </SPAN><SPAN 
class=d2h_breadcrumbsContentsOnly>New Features in This Release</SPAN></DIV>
<H4><A name=_Toc242498268></A><A name=_Toc283129527>SNP Viewer</A></H4>
<P class=MsoNormal>The SNP Viewer allows you to analyze individual base 
variations in DNA sequences for normal and tumor tissue samples in a SNP 
array.&nbsp; The viewer supports both copy number analysis and loss of 
heterozygosity (LOH) analysis.</P>
<P class=MsoNormal>SNP array data, such as Affymetrix Genome-Wide Human SNP 
Array, is required for the SNP Viewer.&nbsp; In this release, the only study 
that contains SNP array data is the public study GSE19539 on ovarian 
cancer.<SPAN class=MsoHyperlink><SPAN style="FONT-SIZE: 9pt"></SPAN></SPAN></P>
<P class=MsoNormal>Dataset Explorer uses the Broad Institute&#8217;s GenePattern 
genomic analysis platform to generate SNP visualizations.&nbsp; For information 
about the Broad Institute&#8217;s SNPViewer, see the document <SPAN 
class=Bold>SNPViewer </SPAN>on the following site:</P>
<P class=MsoNormal><A 
href="http://www.broadinstitute.org/cgi-bin/cancer/software/genepattern/modules/gp_modules.cgi"><SPAN 
style="FONT-SIZE: 9pt">http://www.broadinstitute.org/cgi-bin/cancer/software/genepattern/modules/gp_modules.cgi</SPAN></A></P>
<P style="tab-stops: 0pt" class=ListNumStart>To run the SNP Viewer:</P>
<P style="tab-stops: 18.0pt 36.0pt 54.0pt 72.0pt 90.0pt 108.0pt" 
class=MsoListNumber>1.&nbsp; Run tranSMART, then click the <SPAN 
class=Bold>Dataset Explorer</SPAN> tab.</P>
<P style="tab-stops: 18.0pt 36.0pt 54.0pt 72.0pt 90.0pt 108.0pt" 
class=MsoListNumber>2.&nbsp; Select the study of of interest.</P>
<P style="tab-stops: 18.0pt 36.0pt 54.0pt 72.0pt 90.0pt 108.0pt" 
class=MsoListNumber>3.&nbsp; Drag the SNP array data into an empty box in one or 
both subsets.</P>
<P style="tab-stops: 18.0pt 36.0pt 54.0pt 72.0pt 90.0pt 108.0pt" 
class=MsoListNumber>4.&nbsp; Define the cohorts whose data points will be 
represented in the SNP visualization.</P>
<P style="tab-stops: 18.0pt 36.0pt 54.0pt 72.0pt 90.0pt 108.0pt" 
class=MsoListNumber>5.&nbsp; Click <SPAN class=Bold>Advanced &gt; 
SNPViewer</SPAN>.</P>
<P style="tab-stops: 18.0pt 36.0pt 54.0pt 72.0pt 90.0pt 108.0pt" 
class=MsoListNumber>6.&nbsp; On the SNPViewer dialog, select one or more 
chromosomes to include in the visualization, or select <SPAN 
class=Bold>ALL</SPAN> to include all chromosomes.</P>
<P style="MARGIN-RIGHT: 24pt" class=Note><B><SPAN 
style="COLOR: #1f497d">Note:</SPAN></B><SPAN style="COLOR: black">&nbsp; For 
performance reasons, select only those chromosomes that are of particular 
interest to you.&nbsp; The dataset for all chromosomes is extremely large.&nbsp; 
</SPAN></P>
<P style="tab-stops: 18.0pt 36.0pt 54.0pt 72.0pt 90.0pt 108.0pt" 
class=MsoListNumber>7.&nbsp; Click <SPAN class=Bold>OK</SPAN> to display the 
visualization.</P>
<P class=MsoNormalIndent><SPAN style="COLOR: black">The visualization may 
require several minutes to display a large dataset, such as when multiple 
chromosomes are selected.</SPAN></P></DIV></DIV></BODY></HTML>
